ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/processor/TransactionTracer.java

```diff
@@ -46,7 +61,71 @@ public TransactionTracer(final BlockReplay blockReplay) {
                   tracer,
                   new BlockHashLookup(header, blockchain),
                   false);
-          return new TransactionTrace(transaction, result, tracer.getTraceFrames());
+          timer.stop();
+          return new TransactionTrace(
+              transaction, result, tracer.getTraceFrames(), timer.elapsed(TimeUnit.NANOSECONDS));
+        });
+  }
+
+  public Optional<TransactionTrace> traceTransaction(
+      final Hash blockHash, final Hash transactionHash, final EVMToolTracer tracer) {
+    return blockReplay.beforeTransactionInBlock(
+        blockHash,
+        transactionHash,
+        (transaction, header, blockchain, mutableWorldState, transactionProcessor) -> {
+          final Stopwatch timer = Stopwatch.createStarted();
+          final Result result =
+              transactionProcessor.processTransaction(
+                  blockchain,
+                  mutableWorldState.updater(),
+                  header,
+                  transaction,
+                  header.getCoinbase(),
+                  tracer,
+                  new BlockHashLookup(header, blockchain),
+                  false,
+                  new TransactionValidationParams.Builder().allowFutureNonce(true).build());
+
+          timer.stop();
+          return new TransactionTrace(
+              transaction, result, new ArrayList<>(), timer.elapsed(TimeUnit.NANOSECONDS));
         });
   }
+
+  public List<String> traceTransactionToFile(
+      final Hash blockHash,
+      final List<Transaction> transactions,
+      final Optional<TransactionTraceParams> transactionTraceParams) {
+    final List<String> traces = new ArrayList<>();
+    try {
+      final Optional<Hash> selectedHash =
+          transactionTraceParams.flatMap(TransactionTraceParams::getTransactionHash);
+      final boolean showMemory =
+          transactionTraceParams.map(TransactionTraceParams::disableMemory).orElse(true);
+      for (int i = 0; i < transactions.size(); i++) {
+        final Transaction transaction = transactions.get(i);
+        if (selectedHash.isEmpty()
+            || selectedHash.filter(isEqual(transaction.getHash())).isPresent()) {
+          final File tmpFile =
```

Comment:
Do we want these going to tmp?  perhaps our data root under traces?

Reply:
I don't have a strong opinion on it. I did this to be consistent with the other clients. But I just changed to put it in the data/traces now